doc/parameter_server.md

```diff
@@ -0,0 +1,229 @@
+2. Pserver gets the key and queries kv store to get corresponding parameter
+3. Pserver calls optimizer, applys <key, gradient> to <key, parameter>
+4. Pserver updates <key, parameter> back to kv store
+5. Pserver saves model checkpoint to disk periodially
+
+
+**Note**
+
+We could implement the kv store by ourselves, or we could use some already solution, such as redis.
+
+## Worker
+
+- Worker defines a forward/backward computation proccess
+- Worker defines a dataloader module
+
+Workflow:
+
+1. Before forward layer starting, worker sends key to pserver, and pull <key, parameter> back from pserver
+2. Worker starts forward layer computation
+3. After backward layer finishing, worker generates <key, gradient>
+4. Worker pushes <key, gradient> to pserver
+
+
+There is also another kind of worker who does an evalution job. It define its own evalution computation process and dataloader of validation data. It pulls <key, parameter> from pserver, but never push data to pserver.
+
+
+**Note**
+
+There could be some same item id in a minibatch data. So some gradient vector of embedding table will have the same item id. We need to sum these gradient before pushing to pserver.
+
+
+## Delayed model updating
+
+In order to reduce the communication overhead between workers and pservers, we propose a stragety called delayed model updating.
+
+A worker runs a round of forward/backward computation using its local model, and keep the gradients in its local. After several rounds finishing, it push gradients to pservers.
+
+**Note**
+
+Since local model is only a part of global model, in some circumstance, worker still has to pull embedding vector parameter from pservers if there exits unknow item ids in a minibatch data.
+
+In async mode, this will lead to relative newer embedding part parameters, but relative older other part parameters.
+
+
+## Elastic scheduling
+
+### Analysis
+
+First, we analyse the hardware resources that pserver and worker consume.
+
+
+| role| CPU | memory | network bandwidth |  disk |
+| :----  |:----  |:---|:---|:---|
+| worker  | high| medium |  medium | low|
+| pserver | low | medium | high| medium |
+
+
+Worker does the forward/backward computation and consumes CPU most, while pserver provides kvstore service and consumes network bandwidth most.
+
+Since our target is to scheduling deep learning jobs elasticly, we must have the ability to increase/decrease CPU resources of worker, and increase/decrease network bandwidth resources of pserver.
+
+### Solution
+
+As we are using asynchronous SGD, increasing/decreasing CPU resources of worker is trivial. We can just create more workers or kill some workers under different cluster status.
+
+For example, if job A requests 50 CPU cores, but current cluster only has 30 CPU cores. The cluster will lanuch the job immediately with 30 CPU cores. If the cluster has more CPU cores after a while, it will tell the master of job A to occupy the idle CPU cores and create more workers.
```

Comment:
Current ElasticDL lets user set an upper limit of the number of workers and tries to launch that number of worker pods by k8s APIs.

Reply:
I mean ElasticDL should also have the ability to create more workers even the job starting for a while,  if hardware resources are added.

So, `someone` is needed to tell master to create more workers, the `someone` could be the cluster scheduler.